threatgrid/ctia
# Add rate limiting to CTIA

Add a rate limiter

> **Epic** threatgrid/iroh#2807

This PR adds a rate limiter by group on all routes. See the modified Readme.md for more details.

When the rate limit is not reached, the following headers are returned in the response:
```
"x-ratelimit-group-limit": "18000",
```

If the rate limit is exceeded:
- The client receives a response with the 429 HTTP status, a `retry-after` header and the JSON message `{"error": "Too Many Requests"}`
- A message is logged with the `:info` level

<a name="qa">[§](#qa)</a> QA
============================

1. Check that Cisco Threat Response client is not rate limited (by checking the response headers)
2. Create an Oauth2 client with CTR and use it with CTIA APIs
3. The response headers `x-ratelimit-group-limit` should be set in the response
4. Make the number of requests to exceed the rate limit and check that future calls are blocked

<a name="ops">[§](#ops)</a> Ops
===============================

- Config change needed: https://github.com/threatgrid/tenzin-config/pull/36

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
public: AMP Global Intel and Private AMP Global Intel APIs are now rate limited
```
